### ISSUE TITLE
`supabase functions serve [--all]` unusable since CLI v1.50.1 (blocked by CORS error on preflight requests)

### ISSUE THREAD

**Issue by vejja** [supabase/cli#1020]
It is not possible to use `supabase functions serve` anymore in a local environment since v1.49.0

All edge functions requests are blocked due to CORS rules.
This is due to the edge-runtime returning `408 - Request timeout` on all preflight `OPTIONS` requests.

Screenshots courtesy of user @joshmarnold

![CleanShot_2023-04-16_at_11 09 32](https://user-images.githubusercontent.com/7295259/233671513-d3c58ed4-9977-4edb-a8a4-c1ad4c4c63dd.png)
![CleanShot_2023-04-15_at_19 53 36](https://user-images.githubusercontent.com/7295259/233671510-d7b6815c-e0eb-478a-9c34-47e4c8140fa9.png)
![CleanShot_2023-04-15_at_19 55 00](https://user-images.githubusercontent.com/7295259/233671506-acc8af8f-9739-46b1-8b47-5719cb1175fd.png)
![CleanShot_2023-04-15_at_19 54 49](https://user-images.githubusercontent.com/7295259/233671500-4ddd6705-929b-462c-8f02-95e1300fbaec.png)

Please see full report on the underlying edge-runtime issue at :
https://github.com/supabase/edge-runtime/issues/50

**Comment by soedirgo** [comment]
Will be fixed once https://github.com/supabase/edge-runtime/pull/56 lands - thanks for reporting!

**Comment by johanneswuerbach** [comment]
Also ran into this when trying to write a supabase edge function for the first time and it definitely caused some frustration (I obviously assumed the problem was on my side).

Additionally when supabase is installed via brew, there isn't an easy way to downgrade your client.

Wouldn't it be possible to pin the cli to an older edge runtime version, until this is released?

**Comment by vejja** [comment]
> Wouldn't it be possible to pin the cli to an older edge runtime version, until this is fixed?

I think that's a very good idea

@soedirgo we know that it was commit 661d0813 that introduced the bug, would it be possible to just revert that one for the CLI?

**Comment by vejja** [comment]
I felt I would add a bit more context here for people to better understand the issue:

**This bug actually prevents any local development of edge functions**

When supabase-js tries to `invoke` an edge function, the browser will always send an `OPTIONS` request first. 
Because that requests errors out with `408`, the browser CORS rules then prohibits the `invoke` call.

To be clear, it is not possible to use edge functions anymore since CLI v.1.49.0

Hence the suggestion to revert the responsible commit.

**Comment by henningko** [comment]
Agreed, this is very frustrating, especially when other bug fixes from newer CLI versions are required to continue development.

**Comment by vejja** [comment]
> Agreed, this is very frustrating, especially when other bug fixes from newer CLI versions are required to continue development.

Yes - to be fair I do suspect that the issue originates deep in the edge-runtime Rust code, which itself is a huge piece of work.
Hence the suggestion to revert the responsible commit, which means not using the edge-runtime at all until it is stabilized.
I understand going back to deno-realy would be a step backwards, but unfortunately I can't see a better alternative for now.

**Comment by sweatybridge** [comment]
I believe it's still possible to use the legacy deno relay by passing in function name to the serve command. For eg.

```bash
supabase functions serve hello_world
```

It's a bit inconvenient but hopefully it unblocks local development for you.

**Comment by vejja** [comment]
> I believe it's still possible to use the legacy deno relay by passing in function name to the serve command. For eg.

Yes ! This is absolutely right.
It does unblock the development process if you have only one function to serve, which is very helpful.

Maybe it would be good mentioning somewhere that `supabase functions serve` without a <function-name> should be avoided at all costs for now?

**Comment by vejja** [comment]
@sweatybridge : would you reconsider reverting to deno-relay temporarily ?
The issue has been opened for over 2 weeks now and there is no easy fix on the edge-runtime.
FYI there have been 2 proposed patches but none are working currently. The issues lies in the interface between the Rust and JS code pieces that transfer the underlying body stream, they are not compatible yet so they can't handle empty bodies for now unfortunately.

**Comment by sweatybridge** [comment]
An alternative is to exclude edge runtime from starting automatically with `-x` flag, ie.

```bash
supabase start -x edge-runtime
```

After that you can serve the function with deno-relay as [mentioned](https://github.com/supabase/cli/issues/1020#issuecomment-1518003563).

I'm not sure if reverting is necessary given the above workaround?

**Comment by vejja** [comment]
> An alternative is to exclude edge runtime from starting automatically with `-x` flag, ie.
> 
> ```shell
> supabase start -x edge-runtime
> ```
> 
> After that you can serve the function with deno-relay as [mentioned](https://github.com/supabase/cli/issues/1020#issuecomment-1518003563).
> 
> I'm not sure if reverting is necessary given the above workaround?

Only works if I want to serve a single function, in which case your workaround works. But running `supabase functions serve [--all]` pulls the edge-runtime docker container and starts it anyway. 

The use case for running `supabase functions serve [--all]` is for people who have split their code across multiple functions - which is my case. I cannot serve my code base anymore due to the bug.
Given that the only purpose of `supabase functions serve [--all]` is to address this use case, I do think that reverting is necessary while the fix is being worked on.

In addition, given that the edge-runtime cannot serve CORS-compliant requests, I don't see the point of maintaining it in the CLI for now ?

**Comment by vejja** [comment]
I am adding additional context here because as I realized I wrongly assumed the issue was related to deno-relay vs edge-runtime, I have done some further research.

1. All CLI versions up to v1.50.0 (included) are working. 
When you run `supabase functions serve [--all]`, you can see the following line being logged in the console:
> Serving **supabase/functions** 

Then all functions are being served correctly 

2. Starting from CLI v1.50.1 (included), it is broken. 
When you run `supabase functions serve [--all]`, you can see two lines being logged in the console:
> Setting up Edge Functions runtime...
> Serving functions on http://localhost:54321/functions/v1/<function-name>

Then whenever you invoke the function, you will see 2 groups of 7 lines in the console:
> serving the request with /home/deno/functions/test
serving the request with /home/deno/functions/test
serving the request with /home/deno/functions/test
serving the request with /home/deno/functions/test
serving the request with /home/deno/functions/test
serving the request with /home/deno/functions/test
serving the request with /home/deno/functions/test
Listening on http://localhost:9999/
Listening on http://localhost:9999/
Listening on http://localhost:9999/
Listening on http://localhost:9999/
Listening on http://localhost:9999/
Listening on http://localhost:9999/
Listening on http://localhost:9999/

The rest of the process is then aborted client-side do the CORS issue on the preflight `OPTIONS` request.

In a nutshell the bug started with v1.50.1, and it is correlated to the change in the way the console reports the functions being served.

**Comment by laktek** [comment]
@vejja Can you try upgrading to Supabase CLI 1.52.3+ to see if you still experience these issues?

**Comment by vejja** [comment]
Problem solved !
Many many thanks

**Comment by cohlar** [comment]
Hi, this is still not working for me with supabase `1.52.3` - I'm on Windows 11, it used to work fine in a prior version (not sure exactly which one, probably before `1.50.1`).
When I serve a specific function, it all works fine.

**Comment by laktek** [comment]
@cohlar There's a newer version of the CLI (1.54.3+), including a loading fix for import maps. Can you try upgrading to the latest CLI?

Also, if the newer update still doesn't work, can you share the `import_map.json` used?

**Comment by cohlar** [comment]
@laktek I just tried with supabase 1.54.3 and the issue remains.

My `import_map.json`:
```json
{
  "imports": {
    "http_server": "https://deno.land/std@0.171.0/http/server.ts?exports=serve",
    "supabase": "https://cdn.skypack.dev/@supabase/supabase-js?dts"
  }
}
```

**Comment by henningko** [comment]
Working again for me with latest version. Thanks for the work!

**Comment by laktek** [comment]
@cohlar If you modify your function to directly use the import URLs (instead of using the import map), does it work?

eg: within your function `import { serve } from "https://deno.land/std@0.171.0/http/server.ts"`

**Comment by johanneswuerbach** [comment]
While 200 CORS seems to work now, I'm still getting an error with `return new Response(null, { headers: corsHeaders, status: 204 })`

```
TypeError: Response with null body status cannot have body
    at initializeAResponse (ext:deno_fetch/23_response.js:202:13)
    at new Response (ext:deno_fetch/23_response.js:337:5)
    at UserWorker.fetch (ext:sb_user_workers/user_workers.js:52:16)
    at async Server.<anonymous> (file:///home/deno/main/index.ts:102:16)
    at async Server.#respond (https://deno.land/std@0.182.0/http/server.ts:220:24)
```

Switching to `return new Response('ok', { headers: corsHeaders, status: 200 })` solved this.

**Comment by laktek** [comment]
@johanneswuerbach This looks like a separate issue. Could you create a GitHub issue for it?

**Comment by cohlar** [comment]
@laktek I just tried after replacing all imports with URL imports and removing `import_map.json`, and the issue remains 😞

**Comment by laktek** [comment]
@cohlar Does this happen only for OPTIONS requests or do you get an error for all requests? Can you share the error you receive?

Also, try doing `supabase stop; supabase start` after updating the CLI. It could be possible it's using an older image.

**Comment by cohlar** [comment]
@laktek sorry for the late reply.
I get CORS errors in the browser, see screenshots below:

![image](https://user-images.githubusercontent.com/57757008/236773374-2efa3b0c-1b38-4dd6-98a0-10afbf1a9c60.png)

![image](https://user-images.githubusercontent.com/57757008/236773447-316f6306-e20f-4bc7-ba8f-baaf7901439c.png)
I do not get any error in the supabase CLI.

My functions have an OPTIONS part to handle CORS (see below) before handling their logic, it feels like that is where they fail.
```js
  if (req.method === 'OPTIONS') {
    return new Response('ok', { headers });
  }
  ```

**Comment by cohlar** [comment]
I just tried with supabase `1.55.1`, the issue remains.

**Comment by laktek** [comment]
@cohlar Can you `console.log(headers)` within your function? Just want to make sure if the headers are loaded correctly.

Alternatively, can you hardcode them into the block that is handling CORS:
```
if (req.method === 'OPTIONS') {
    const headers = {
       'Access-Control-Allow-Origin': '*',
       'Access-Control-Allow-Headers': 'authorization, x-client-info, apikey, content-type',
     }
     return new Response('ok', { headers });
 }
```

If that still doesn't solve the problem, can you share how you call the function from the browser? Are you using `supabase-js`?
